enhancements/scheduling/prevent-user-workloads-from-being-scheduled-on-control-plane-nodes.md

```diff
@@ -0,0 +1,526 @@
+  This coordination ensures that the designated workloads are appropriately scheduled on the protected nodes.
+
+#### Workflow Steps
+
+1. Node tainting:
+
+- Cluster Administrator applies NoExecute taints to control plane nodes and any other special node groups deemed necessary to protect.
+  This foundational step prevents unauthorized workloads from being scheduled (or directly placed with spec.nodeName field) on these critical nodes.
+
+```
+oc taint nodes ip-XX-XX-XX-XXX.ec2.internal node-role.kubernetes.io/control-plane:NoExecute-
+```
+
+2. Validating Admission Policy provided:
+
+- As part of this proposal a `ValidatingAdmissionPolicy` will be provided as part of the static manifests applied by Kube Scheduler Operator ([staticresourcecontroller](https://github.com/openshift/cluster-kube-scheduler-operator/blob/67304344622e50cbcd95e7a05056294e9d96df99/pkg/operator/starter.go#L108)).
+
+```
+apiVersion: admissionregistration.k8s.io/v1beta1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "control-plane-scheduling-policy"
+spec:
+  failurePolicy: Fail
+  matchConstraints:
+    resourceRules:
+    - apiGroups:   [""]
+      apiVersions: ["v1"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["pods"]
+  validations:
+    - expression: >
+        (request.userInfo.groups.exists(g, g == 'system:nodes') &&
+        "kubernetes.io/config.mirror" in object.metadata.annotations) ||
+        object.spec.tolerations.all(toleration,
+          toleration.effect != 'NoExecute' ||
+          authorizer.subjectAccessReview(object.metadata.namespace, object.spec.serviceAccountName)
+            .group('scheduling.openshift.io')
+            .resource(toleration.key)
+            .subresource('toleration')
+            .name(toleration.operator == 'Exists' ? '*' : toleration.value)
+            .verb('Allow' + toleration.effect)
+            .allowed())
+      messageExpression: >
+        "Pod toleration for 'NoExecute' is not authorized for service account '" + object.spec.serviceAccountName + "' in namespace '" + object.metadata.namespace + "'."
+
+```
+
+- Note that as part of the expression we also exclude static/mirror pods submitted by kubelet (group 'system:node'). If the static pod manifest is on the node, the pod will run regardless, so there is no point in validating the permissions there, but we need to make sure that is the case by checking the group, and accept this exception.
+
+3. Validating Admission Policy Binding:
+
+- As part of this approach a ValidatigAdmissionPolicyBinding is also provided. The Cluster Administrator updates `validationActions` field to flip the switch and enable actual enforcement. This is the only field where we allow change, but admins can also create their own VAP/VAPB process if they need to. Binding shown here affects all namespaces.
```

Comment:
See discussion: https://github.com/openshift/enhancements/pull/1583#discussion_r1607263991

We can revisit this, of course. Idea here was to keep it simple. Customers can actually already achieve everything described in this EP themselves out of the box, but doing steps manually. We are just providing a standardized way of enabling this to make supporting this easier. 

Also this discussion is relevant: https://redhat-internal.slack.com/archives/CC3CZCQHM/p1713811003793789?thread_ts=1713794008.195589&cid=CC3CZCQHM

Reply:
I've skimmed both places and none of the two explicitly moves the responsibility to the cluster admin. I think the intention is to have a high level API to express these options, but the underlying resources are managed and configured by the operator. The majority of our API, if not all is designed with that in mind, to hide the complexity from cluster-admins.

Reply:
It's not totally explicit, but we moved from the previous approach to avoid having API changes and a new controller or new admission plugin. With that in mind we don't have a place to a new field or a new API to control this, if the whole idea is to take advantage of existing APIs and existing functionality of the Scheduler, API Server and Kubelet. (Only providing VAP + VAPBs to end users and admins).

That's why VAP with cell expressions was created, right? So we can let admins have their own rules implemented with CEL expressions and at most we can provide default or standard VAPs when needed (with no new controllers or plugins or new APIs).

Reply:
@deads2k would you please chime in? Please let me know if I had the wrong interpretation here.